File: backend/src/BeOffline.Api/Services/Names.cs
namespace BeOffline.Api.Services;

/// <summary>
/// People are named by first name in everything the user actually reads.
///
/// A Google display name is usually someone's full legal name, which is both
/// more than a notification line can fit and more than you need to know who is
/// asking to open Instagram. The full name stays in the DTOs — the client still
/// shows it in group member management, where telling two people apart is the
/// entire job.
/// </summary>
public static class Names
{
    /// <summary>First word of <paramref name="displayName"/>, or <paramref name="fallback"/>.</summary>
    public static string First(string? displayName, string fallback = "Your partner")
    {
        if (string.IsNullOrWhiteSpace(displayName)) return fallback;
        var trimmed = displayName.Trim();
        var space = trimmed.IndexOf(' ');
        var first = space < 0 ? trimmed : trimmed[..space];
        return string.IsNullOrEmpty(first) ? fallback : first;
    }
}

File: app/src/main/java/com/beoffline/app/util/Names.kt
package com.beoffline.app.util

/**
 * People are shown by first name. A Google account's display name is usually
 * the person's full legal name, which is both more than you need to know who
 * approved your unlock and more than fits in a notification line.
 *
 * This is a display rule only — the full name stays in the cache and in the
 * member list, where telling two people apart actually matters.
 */
fun String?.firstName(fallback: String = "Your partner"): String {
    val trimmed = this?.trim().orEmpty()
    if (trimmed.isEmpty()) return fallback
    return trimmed.substringBefore(' ').ifEmpty { fallback }
}

/**
 * First names for a set of people, keeping enough of the full name to tell
 * two Priyas apart. Anyone whose first name is unique in [fullNames] is shown
 * by first name alone; the rest keep a surname initial ("Priya S.").
 */
fun disambiguatedFirstNames(fullNames: List<String>): Map<String, String> {
    val byFirst = fullNames.groupBy { it.firstName(fallback = it) }
    return fullNames.associateWith { full ->
        val first = full.firstName(fallback = full)
        val sharers = byFirst[first].orEmpty()
        if (sharers.size <= 1) {
            first
        } else {
            val rest = full.trim().substringAfter(' ', "").trim()
            if (rest.isEmpty()) first else "$first ${rest.first().uppercaseChar()}."
        }
    }
}
